dev-tools/mage/integtest.go

```diff
@@ -186,53 +218,175 @@ func runInIntegTestEnv(mageTarget string, test func() error, passThroughEnvVars
+			args...,
+		)
+		if err != nil {
+			return err
+		}
+	}
+	if testEnv.HasType(TestingEnvKubernetes) {
+		clusterName := kindClusterName()
+		stdOut := ioutil.Discard
+		stdErr := ioutil.Discard
+		if mg.Verbose() {
+			stdOut = os.Stdout
+			stdErr = os.Stderr
+		}
+
+		kubeConfig := os.Getenv("KUBECONFIG")
+		if kubeConfig == "" {
+			// Create a kubernetes cluster with kind.
+			if mg.Verbose() {
+				fmt.Println(">> Starting Kind test environment...")
+			}
+
+			kubeCfgDir := filepath.Join("build", "kind", clusterName)
+			kubeCfgDir, err = filepath.Abs(kubeCfgDir)
+			if err != nil {
+				return err
+			}
+			kubeConfig = filepath.Join(kubeCfgDir, "kubecfg")
+			if err := os.MkdirAll(kubeCfgDir, os.ModePerm); err != nil {
+				return err
+			}
+
+			args := []string{
+				"create",
+				"cluster",
+				"--name",
+				kindClusterName(),
+				"--kubeconfig", kubeConfig,
+				"--wait",
+				"300s",
+			}
+			kubeVersion := os.Getenv("K8S_VERSION")
+			if kubeVersion != "" {
+				args = append(args, "--image", fmt.Sprintf("kindest/node:%s", kubeVersion))
+			}
+
+			_, err = sh.Exec(
+				map[string]string{},
+				stdOut,
+				stdErr,
+				"kind",
+				args...,
+			)
+			if err != nil {
+				return err
+			}
+		}
+
+		manifestPath, _ := testEnv.GetTypeSource(TestingEnvKubernetes)
+		kubeEnv, err := integTestKindEnvVars(kubeConfig)
+		if err != nil {
+			return err
+		}
+
+		if mg.Verbose() {
+			fmt.Println(">> Applying module manifest to cluster...")
+		}
+
+		// Apply the manifest from the module. Module uses the manifest as the
+		// base for running inside the cluster.
+		if err := KubectlApply(kubeEnv, stdOut, stdErr, manifestPath); err != nil {
+			return errors.Wrapf(err, "failed to apply manifest %s", manifestPath)
+		}
+		defer func() {
+			if mg.Verbose() {
+				fmt.Println(">> Deleting module manifest from cluster...")
+			}
+			if err := KubectlDelete(kubeEnv, stdOut, stdErr, manifestPath); err != nil {
+				log.Printf("%s", errors.Wrapf(err, "failed to apply manifest %s", manifestPath))
+			}
+		}()
+
+		// Execute the test inside of kubernetes.
+		remoteEnv := map[string]string{
+			beatsDockerIntegrationTestEnvVar: "true",
+			"BEAT_STRICT_PERMS":              "false",
+		}
+		if UseVendor {
+			remoteEnv["GOFLAGS"] = "-mod=vendor"
+		}
+		for _, envVar := range passThroughEnvVars {
+			remoteEnv[envVar] = os.Getenv(envVar)
+		}
+		if mg.Verbose() {
+			remoteEnv["MAGEFILE_VERBOSE"] = "1"
+		}
+		if mg.Verbose() {
+			fmt.Println(">> Executing tests inside of cluster...")
+		}
+
+		destDir := filepath.Join("/go/src", repo.CanonicalRootImportPath)
+		workDir := filepath.Join(destDir, repo.SubDir)
+		remote, err := NewKubeRemote(kubeConfig, "default", clusterName, workDir, destDir, repo.RootDir)
+		if err != nil {
+			return err
+		}
+		// Uses `os.Stdout` directly as its output should always be shown.
+		err = remote.Run(remoteEnv, os.Stdout, stdErr, magePath, mageTarget)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
```

Comment:
Split this in different methods please, at least one for docker compose and another one for kubernetes.

Reply:
I have completely separated this code and converted everything into different pieces.

See `DockerIntegrationTester` and `KubernetesIntegrationTester` in latest commit.